### PR TITLE
Edge cases

### PR DESCRIPTION
**Pull request by jifeon**
It does not affect current `Parser#parse` method a lot, but adds mechanisms for edge cases, which need to be tested thoroughly. 

For review @maZahaca

### CODE REVIEW

File: docker/index.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ parser
             console.log('Execution time: ' + ((new Date).getTime() - time));
             console.log('Results:');
         }
-        console.log(results);
+        console.log(require('util').inspect(results, {showHidden: false, depth: null}));
```
Comment (maZahaca) [inline docker/index.js]: what's "inspect" for ?

File: lib/Parser.js [hunk 2]
```diff
@@ -169,22 +170,117 @@ Parser.prototype = {
...
+                    return results;
+                },
+                e => {
+                    debug('Saving partial results because of error %o', e);
+                    this._partialResults = results;
+                    throw e;
+                });
+    },
+
+    moveYourFeet(stages) {
```
Comment (maZahaca) [inline lib/Parser.js]: wtf ?

File: lib/Parser.js [hunk 3]
```diff
@@ -169,22 +170,117 @@ Parser.prototype = {
...
+
+        return this._env
+            .prepare()
+            .then(() => milestones.reduce((promise, milestone) => {
+                return promise.then(() => this.passMilestone(milestone, edgeCases));
+            }, Promise.resolve()))
+            .then(
+                () => this._env.tearDown(),
+                e => this._env
+                    .snapshot('error')
```
Comment (maZahaca) [inline lib/Parser.js]: this is not a good idea to do it all time, it's quite lazy operation.
Let's use it, if we have special option in phantomenv, we already have one

File: lib/Parser.js [hunk 4]
```diff
@@ -383,7 +477,7 @@ Parser.prototype = {
         const format = results => {
             if (_.contains([this.TYPES.SIMPLE, this.TYPES.GET, this.TYPES.VALUE, this.TYPES.COLLECTION], ruleType)) {
                 if (Array.isArray(results) && rule.type !== 'array') {
-                    return results.join(rule.separator || ' ');
+                    return results.length === 1 ? results[0] : results.join(rule.separator || ' ');
```
Comment (maZahaca) [inline lib/Parser.js]: one more trick in goose logic

File: lib/Parser.js [hunk 5]
```diff
@@ -632,24 +730,28 @@ Parser.prototype = {
...
             return nodes.map(function(node) {
                 switch (filter.type) {
-                    case 'attr':
-                        return node.getAttribute(filter.value);
                     case 'child':
                         var childNode = node.childNodes[filter.value];
                         return childNode.textContent;
+                    case 'attr':
+                        return node.getAttribute(filter.value);
+                    case 'prop':
```
Comment (maZahaca) [inline lib/Parser.js]: could you explain me one more time what is the difference between attribute and property?

**Comment by maZahaca** [comment]
Go ahead!

File: lib/Parser.js [hunk 6]
```diff
@@ -169,22 +170,117 @@ Parser.prototype = {
...
+                },
+                e => {
+                    debug('Saving partial results because of error %o', e);
+                    this._partialResults = results;
+                    throw e;
+                });
+    },
+
+    moveYourFeet(stages) {
+        debug('Hit the road!');
```
Comment (jifeon) [inline lib/Parser.js]: and never come back

File: lib/Parser.js (hunk 2 above)
Comment (jifeon) [inline lib/Parser.js]: The Goose, like a hero, keeps his own way, with his own mistakes, ups and downs, moving his little paws.

File: lib/Parser.js (hunk 3 above)
Comment (jifeon) [inline lib/Parser.js]: agree

File: lib/Parser.js (hunk 4 above)
Comment (jifeon) [inline lib/Parser.js]: yeah, it's for booleans if you do not parse an array: `[false].join(' ') === 'false'`

Comment (jifeon) [inline lib/Parser.js]: tests are on the way

File: lib/Parser.js (hunk 5 above)
Comment (jifeon) [inline lib/Parser.js]: http://stackoverflow.com/questions/6003819/properties-and-attributes-in-html

File: docker/index.js (hunk 1 above)
Comment (jifeon) [inline docker/index.js]:
```
Andrews-MBP:Volumes jifeon$ node
> console.log({a: {b: {c: 2, d: [1,2,3,4,5,6,7,8,9,0]}}})
{ a: { b: { c: 2, d: [Object] } } }
undefined
> require('util').inspect({a: {b: {c: 2, d: [1,2,3,4,5,6,7,8,9,0]}}}, {depth: null})
'{ a: { b: { c: 2, d: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 0 ] } } }'
```

**Pull request merged**